File: StriverSheet/backtracking/Nqueen.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution
{
public:
    int n;
    vector<vector<string>> ans;
    vector<string> board;

    // ---------------- SAFETY ----------------
    bool isSafe(int row,int col)
    {
        // Column check
        for(int i=0;i<row;i++)
        {
            if(board[i][col]=='Q')
                return false;
        }

        // Left diagonal ↖
        int i=row-1;
        int j=col-1;

        while(i>=0 && j>=0)
        {
            if(board[i][j]=='Q')
                return false;

            i--;
            j--;
        }

        // Right diagonal ↗
        i=row-1;
        j=col+1;

        while(i>=0 && j<n)
        {
            if(board[i][j]=='Q')
                return false;

            i--;//up 
            j++;//down  i.e up +down
        }

        return true;
    }

    // ---------------- BACKTRACK ----------------
    void solve(int row)
    {
        if(row==n)
        {
            ans.push_back(board);
            return;
        }

        for(int col=0;col<n;col++)
        {
            if(isSafe(row,col))
            {
                board[row][col]='Q';   // place

                solve(row+1);          // recurse

                board[row][col]='.';   // backtrack
            }
        }
    }

    // ---------------- DRIVER ----------------
    vector<vector<string>> solveNQueen(int N)
    {
        n=N;

        board = vector<string>(n, string(n,'.'));

        solve(0);

        return ans;
    }
};

// ---------------- MAIN -------------------
int main()
{
    Solution obj;
    int n=4;

    vector<vector<string>> res=obj.solveNQueen(n);

    for(auto &board:res)
    {
        for(auto &row:board)
            cout<<row<<endl;

        cout<<"---------"<<endl;
    }
}

//here we have to make a 1d string and storing string in each string=row

//"...."
// "...."
// "...."
// "...."


// ['.','.','.','.']

//board[i][j] there i will give the string and j will give the index

//array of string we have to create not an array of array

File: StriverSheet/backtracking/Nqueen.js
let n;
let board;
let ans;
function isSafe(row,col)
{
   //column check 
   for(let i=0;i<row;i++)
   {
    if(board[i][col]=='Q')
    {
        return false;
    }
   }
   //left diagonal check

 for (let i=row-1, j=col-1;
     i>=0 && j>=0;
     i--, j--)
{
   if(board[i][j] === 'Q')
       return false;
}
   //rigth diagonal check
for(let i=row-1, j=col+1;
    i>=0 && j<n;
    i--, j++)
{
    if(board[i][j] === 'Q')
        return false;
}
return true;
}

function solve(row)
{
    if(row==n)
    {
        ans.push(board.map(r=>r.join("")));
        return;
    }
    //fix the row try all column 
    for(let col=0;col<n;col++)
    {
        if(isSafe(row,col))
        {
            board[row][col]='Q';
            solve(row+1);
            board[row][col]='.';
        }
    }
}
function solveNQueen(N)
{
    n=N;
    board=Array(n).fill().map(()=>Array(n).fill('.'));
    ans=[];
    solve(0);
    return ans;
}
console.log(solveNQueen(4));